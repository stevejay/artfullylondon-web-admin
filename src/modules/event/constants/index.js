import eventType from '_src/domain/types/event-type'
import bookingType from '_src/domain/types/booking-type'
import costType from '_src/domain/types/cost-type'
import occurrenceType from '_src/domain/types/occurrence-type'

export const EDIT_EVENT_BASICS_FORM_NAME = 'EditEventBasics'
export const EDIT_EVENT_IMAGES_FORM_NAME = 'EditEventImages'
export const EDIT_EVENT_TAGS_FORM_NAME = 'EditEventTags'
export const EDIT_EVENT_TALENTS_FORM_NAME = 'EditEventTalents'
export const EDIT_EVENT_TIMES_FORM_NAME = 'EditEventTimes'

export const CREATE_TALENT_FORM_NAME = 'CreateTalent'
export const TALENT_MATCHES_FOUND_FORM_NAME = 'TalentMatchesFound'

export const EVENT_TYPE_DROPDOWN_OPTIONS = [
  { value: eventType.PERFORMANCE, label: eventType.PERFORMANCE },
  { value: eventType.EXHIBITION, label: eventType.EXHIBITION }
]

export const COST_TYPE_DROPDOWN_OPTIONS = [
  { value: costType.FREE, label: costType.FREE },
  { value: costType.PAID, label: costType.PAID },
  { value: costType.UNKNOWN, label: costType.UNKNOWN }
]

export const BOOKING_TYPE_DROPDOWN_OPTIONS = [
  { value: bookingType.NOT_REQUIRED, label: 'Not Required' },
  { value: bookingType.REQUIRED, label: 'Required' },
  {
    value: bookingType.REQUIRED_FOR_NON_MEMBERS,
    label: 'Required for Non-Members'
  }
]

export const PERFORMANCE_OCCURRENCE_TYPE_DROPDOWN_OPTIONS = [
  { value: occurrenceType.BOUNDED, label: occurrenceType.BOUNDED },
  { value: occurrenceType.CONTINUOUS, label: occurrenceType.CONTINUOUS },
  { value: occurrenceType.ONETIME, label: 'One-time' }
]

export const EXHIBITION_OCCURRENCE_TYPE_DROPDOWN_OPTIONS = [
  { value: occurrenceType.BOUNDED, label: occurrenceType.BOUNDED },
  { value: occurrenceType.CONTINUOUS, label: occurrenceType.CONTINUOUS }
]

export const RATING_DROPDOWN_OPTIONS = [
  { value: '1', label: 'One star' },
  { value: '2', label: 'Two stars' },
  { value: '3', label: 'Three stars' },
  { value: '4', label: 'Four stars' },
  { value: '5', label: 'Five stars' }
]

export const DURATION_DROPDOWN_OPTIONS = [
  { value: '00:15', label: '15 mins' },
  { value: '00:30', label: '30 mins' },
  { value: '00:45', label: '45 mins' },
  { value: '01:00', label: '1 hour' },
  { value: '01:15', label: '1 hour 15 mins' },
  { value: '01:20', label: '1 hour 20 mins' },
  { value: '01:30', label: '1 hour 30 mins' },
  { value: '01:40', label: '1 hour 40 mins' },
  { value: '01:45', label: '1 hour 45 mins' },
  { value: '02:00', label: '2 hours' },
  { value: '02:15', label: '2 hours 15 mins' },
  { value: '02:20', label: '2 hours 20 mins' },
  { value: '02:30', label: '2 hours 30 mins' },
  { value: '02:40', label: '2 hours 40 mins' },
  { value: '02:45', label: '2 hours 45 mins' },
  { value: '03:00', label: '3 hours' },
  { value: '03:15', label: '3 hours 15 mins' },
  { value: '03:20', label: '3 hours 20 mins' },
  { value: '03:30', label: '3 hours 30 mins' },
  { value: '03:40', label: '3 hours 40 mins' },
  { value: '03:45', label: '3 hours 45 mins' },
  { value: '04:00', label: '4 hours' },
  { value: '05:00', label: '5 hours' },
  { value: '06:00', label: '6 hours' },
  { value: '07:00', label: '7 hours' },
  { value: '08:00', label: '8 hours' }
]

export const MIN_AGE_DROPDOWN_OPTIONS = [
  { value: '', label: 'None' },
  { value: '1', label: '1+' },
  { value: '2', label: '2+' },
  { value: '3', label: '3+' },
  { value: '4', label: '4+' },
  { value: '5', label: '5+' },
  { value: '6', label: '6+' },
  { value: '7', label: '7+' },
  { value: '8', label: '8+' },
  { value: '9', label: '9+' },
  { value: '10', label: '10+' },
  { value: '11', label: '11+' },
  { value: '12', label: '12+' },
  { value: '13', label: '13+' },
  { value: '14', label: '14+' },
  { value: '15', label: '15+' },
  { value: '16', label: '16+' },
  { value: '17', label: '17+' },
  { value: '18', label: '18+' },
  { value: '21', label: '21+' }
]
