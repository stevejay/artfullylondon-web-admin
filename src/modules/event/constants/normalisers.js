export const BASIC_NORMALISER = {
  name: {
    trim: true
  },
  duration: {
    trim: true
  },
  venueGuidance: {
    trim: true
  },
  summary: {
    trim: true
  },
  descriptionCredit: {
    trim: true
  },
  weSay: {
    trim: true
  }
}

export const ALL_NORMALISER = BASIC_NORMALISER
