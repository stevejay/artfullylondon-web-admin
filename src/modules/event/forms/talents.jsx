import React from 'react'
export default () => <div>Talents</div>
