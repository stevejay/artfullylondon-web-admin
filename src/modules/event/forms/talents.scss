.last-divider {
  margin-top: 1.5rem;
}

.modal-form {
  margin-left: 1.5rem;
  margin-right: 1.5rem;
}
