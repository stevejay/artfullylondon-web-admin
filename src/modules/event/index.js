import EventEditOrCreate from './components/event-edit-or-create'
import EventDetail from './components/event-detail'
import { moduleName, reducer, selectors } from './reducers'
import sagas from './sagas'
export { EventEditOrCreate, EventDetail, moduleName, reducer, selectors, sagas }
