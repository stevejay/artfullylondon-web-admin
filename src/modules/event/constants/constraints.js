import _ from 'lodash'
import * as eventConstants from '_src/constants/event'

const DATE_REGEX = /^[12]\d\d\d\/[01]\d\/[0123]\d$|^$/
const MONEY_REGEX = /^[0-9]+(?:\.[0-9]{1,2})?$|^$/
const TIME_REGEX = /^(?:[01][0-9]|2[0-3]):[0-5][0-9]$|^$/

export const BASIC_CONSTRAINT = {
  name: {
    presence: { disallowEmpty: true },
    length: { maximum: 200 }
  },
  eventType: {
    presence: { disallowEmpty: true },
    inclusion: { within: eventConstants.ALLOWED_EVENT_TYPES }
  },
  occurrenceType: {
    presence: { disallowEmpty: true },
    dependency: [
      {
        test: (_, attrs) =>
          attrs.eventType === eventConstants.EVENT_TYPE_PERFORMANCE,
        ensure: value =>
          _.includes(
            eventConstants.ALLOWED_PERFORMANCE_OCCURRENCE_TYPES,
            value
          ),
        message: 'Occurrence Type is not valid for the event type'
      },
      {
        test: (_, attrs) =>
          attrs.eventType === eventConstants.EVENT_TYPE_EXHIBITION,
        ensure: value =>
          _.includes(eventConstants.ALLOWED_EXHIBITION_OCCURRENCE_TYPES, value),
        message: 'Occurrence Type is not valid for the event type'
      }
    ]
  },
  dateFrom: {
    string: true,
    format: { pattern: DATE_REGEX },
    dependency: [
      {
        test: (_, attrs) =>
          attrs.occurrenceType === eventConstants.OCCURRENCE_TYPE_ONETIME ||
          attrs.occurrenceType === eventConstants.OCCURRENCE_TYPE_BOUNDED,
        ensure: value => !!value,
        message: 'From date is required'
      },
      {
        test: (_, attrs) =>
          attrs.occurrenceType === eventConstants.OCCURRENCE_TYPE_CONTINUOUS,
        ensure: value => !value,
        message: 'From date is not allowed here'
      }
    ]
  },
  dateTo: {
    string: true,
    format: { pattern: DATE_REGEX },
    dependency: [
      {
        test: (_, attrs) =>
          attrs.occurrenceType === eventConstants.OCCURRENCE_TYPE_ONETIME ||
          attrs.occurrenceType === eventConstants.OCCURRENCE_TYPE_BOUNDED,
        ensure: value => !!value,
        message: 'To date is required'
      },
      {
        test: (_, attrs) =>
          attrs.occurrenceType === eventConstants.OCCURRENCE_TYPE_CONTINUOUS,
        ensure: value => !value,
        message: 'To date is not allowed here'
      },
      {
        test: (_, attrs) =>
          attrs.dateFrom !== '' &&
          attrs.occurrenceType === eventConstants.OCCURRENCE_TYPE_ONETIME,
        ensure: (value, attrs) => value === attrs.dateFrom,
        message: 'To date must equal From date'
      },
      {
        test: (_, attrs) =>
          attrs.dateFrom !== '' &&
          attrs.occurrenceType === eventConstants.OCCURRENCE_TYPE_BOUNDED,
        ensure: (value, attrs) => value >= attrs.dateFrom,
        message: 'To date must be greater than or equal to From date'
      }
    ]
  },
  costType: {
    presence: { disallowEmpty: true },
    inclusion: { within: eventConstants.ALLOWED_COST_TYPES }
  },
  costFrom: {
    length: { maximum: 6 },
    format: { pattern: MONEY_REGEX },
    dependency: [
      {
        test: (_, attrs) => attrs.costType === eventConstants.COST_TYPE_PAID,
        ensure: value => value >= 0,
        message: 'Cost From is required'
      },
      {
        test: (_, attrs) =>
          attrs.costType === eventConstants.COST_TYPE_FREE ||
          attrs.costType === eventConstants.COST_TYPE_UNKNOWN,
        ensure: value => !value,
        message: 'Cost From is not allowed here'
      }
    ]
  },
  costTo: {
    length: { maximum: 6 },
    format: { pattern: MONEY_REGEX },
    dependency: [
      {
        test: (_, attrs) => attrs.costType === eventConstants.COST_TYPE_PAID,
        ensure: value => value >= 0,
        message: 'Cost To is required'
      },
      {
        test: (_, attrs) =>
          attrs.costType === eventConstants.COST_TYPE_FREE ||
          attrs.costType === eventConstants.COST_TYPE_UNKNOWN,
        ensure: value => !value,
        message: 'Cost To is not allowed here'
      },
      {
        test: (_, attrs) =>
          attrs.costFrom !== '' &&
          attrs.costType === eventConstants.COST_TYPE_PAID,
        ensure: (value, attrs) => Number(value) >= Number(attrs.costFrom),
        message: 'Cost To must be equal to or greater than Cost From'
      }
    ]
  },
  duration: {
    format: { pattern: TIME_REGEX }
  },
  venueGuidance: {
    length: { maximum: 200 }
  },
  summary: {
    presence: { disallowEmpty: true },
    length: { maximum: 140 }
  },
  descriptionCredit: {
    length: { maximum: 200 }
  },
  bookingType: {
    presence: { disallowEmpty: true },
    inclusion: { within: eventConstants.ALLOWED_BOOKING_TYPES }
  },
  bookingOpens: {
    format: { pattern: DATE_REGEX }
  },
  links: {
    length: { maximum: 4, tooLong: 'has too many elements' }
  },
  weSay: {
    length: { maximum: 300 }
  },
  rating: {
    presence: { disallowEmpty: true },
    format: /^\d$/,
    numericality: { greaterThanOrEqualTo: 1, lessThanOrEqualTo: 5 }
  }
}

export const TAG_CONSTRAINT = {
  mediumTags: { presence: true }
}

const TIMES_RANGE_DEPENDENCY = {
  test: (value, attrs) =>
    value && value.length && attrs.timesRanges && attrs.timesRanges.length,
  ensure: (value, attrs) =>
    _.every(
      value,
      element =>
        element.timesRangeId &&
        _.find(
          attrs.timesRanges,
          timesRange => timesRange.id === element.timesRangeId
        )
    )
}

export const TIMES_CONSTRAINT = {
  openingTimes: {
    dependency: {
      ...TIMES_RANGE_DEPENDENCY,
      message: 'All Opening Times must have a times range that exists'
    }
  },
  performances: {
    dependency: {
      ...TIMES_RANGE_DEPENDENCY,
      message: 'All Performances must have a times range that exists'
    }
  }

  // if there are times ranges, all opening times/performances must have times range id.
  /// / if there are no times ranges, all opening times/performances must have
}

export const TALENT_CONSTRAINT = {
  // TODO All talents must have non-null roles
}

export const IMAGE_CONSTRAINT = {
  // no-op
}
