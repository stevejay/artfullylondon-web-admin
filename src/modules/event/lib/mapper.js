import statusType from '_src/domain/types/status-type'
import { entityMapper } from '_src/modules/entity'
import * as eventLib from '_src/shared/lib/event'

export function getInitialValues (event) {
  const isPaid = eventLib.eventIsPaid(event.costType)
  const bookingIsRequired = eventLib.bookingRequired(event.bookingType)
  const hasOccurrenceRange = eventLib.occurrenceHasDateRange(
    event.occurrenceType
  )

  return {
    id: event.id || null,
    status: event.status || statusType.ACTIVE,
    validStatuses: entityMapper.getValidStatusesInitialValue(event.status),
    eventType: event.eventType || '',
    name: event.name || '',
    occurrenceType: event.occurrenceType || '',
    dateFrom: hasOccurrenceRange ? event.dateFrom : null,
    dateTo: hasOccurrenceRange ? event.dateTo : null,
    soldOut: !!event.soldOut,
    costType: event.costType || '',
    costFrom: isPaid ? event.costFrom.toFixed(2) : '',
    costTo: isPaid ? event.costTo.toFixed(2) : '',
    bookingType: event.bookingType || '',
    bookingOpens: bookingIsRequired ? event.bookingOpens : null,
    summary: event.summary || '',
    rating: event.rating ? event.rating.toString() : '3',
    minAge: event.minAge ? event.minAge.toString() : '',
    maxAge: event.maxAge ? event.maxAge.toString() : '',
    eventSeries: event.eventSeries
      ? {
        entityType: event.eventSeries.entityType,
        eventSeriesType: event.eventSeries.eventSeriesType,
        id: event.eventSeries.id,
        name: event.eventSeries.name
      }
      : {},
    venue: event.venue
      ? {
        entityType: event.venue.entityType,
        status: event.venue.status,
        venueType: event.venue.venueType,
        id: event.venue.id,
        name: event.venue.name,
        address: event.venue.address,
        postcode: event.venue.postcode
      }
      : {},
    venueGuidance: event.venueGuidance || '',
    useVenueOpeningTimes: !!event.useVenueOpeningTimes,
    timesRanges: entityMapper.getTimesRangesInitialValue(event.timesRanges),
    openingTimes: entityMapper.getOpeningTimesInitialValue(event.openingTimes),
    additionalOpeningTimes: entityMapper.getAdditionalOpeningTimesInitialValue(
      event.additionalOpeningTimes
    ),
    specialOpeningTimes: entityMapper.getSpecialOpeningTimesInitialValue(
      event.specialOpeningTimes
    ),
    openingTimesClosures: entityMapper.getOpeningTimesClosuresInitialValue(
      event.openingTimesClosures
    ),
    performances: entityMapper.getPerformancesInitialValue(event.performances),
    additionalPerformances: entityMapper.getAdditionalPerformancesInitialValue(
      event.additionalPerformances
    ),
    specialPerformances: entityMapper.getSpecialPerformancesInitialValue(
      event.specialPerformances
    ),
    performancesClosures: entityMapper.getPerformancesClosuresInitialValue(
      event.performancesClosures
    ),
    duration: event.duration || '',
    talents: _getTalentsInitialValue(event.talents),
    audienceTags: event.audienceTags || [],
    mediumTags: event.mediumTags || [],
    styleTags: event.styleTags || [],
    geoTags: event.geoTags || [],
    reviews: event.reviews || [],
    soldOutPerformances: entityMapper.getSoldOutPerformancesInitialValue(
      event.soldOutPerformances
    ),
    description: entityMapper.getRichTextInitialValue(event.description),
    descriptionCredit: event.descriptionCredit || '',
    links: entityMapper.getLinksInitialValue(event.links),
    images: entityMapper.getImagesInitialValue(event.images),
    weSay: event.weSay || '',
    version: event.version || 0
  }
}

// TODO move to the talent/entity module?
function _getTalentsInitialValue (talents) {
  return (talents || []).map(talent => ({
    ...talent,
    key: talent.id,
    roles: (talent.roles || []).join(', '),
    characters: (talent.characters || []).join(', ')
  }))
}

export function mapSubmittedValues (values) {
  const isFree = !eventLib.eventIsPaid(values.costType)
  const costFrom = isFree ? null : parseFloat(values.costFrom)
  const costTo = isFree ? null : parseFloat(values.costTo)

  const bookingNotRequired = !eventLib.bookingRequired(values.bookingType)
  const bookingOpens = bookingNotRequired ? null : values.bookingOpens
  const hasOccurrenceRange = eventLib.occurrenceHasDateRange(
    values.occurrenceType
  )

  return {
    status: values.status,
    name: values.name.trim(),
    eventType: values.eventType,
    occurrenceType: values.occurrenceType,
    dateFrom: hasOccurrenceRange ? values.dateFrom : null,
    dateTo: hasOccurrenceRange ? values.dateTo : null,
    summary: values.summary.trim(),
    costType: values.costType,
    costFrom: costFrom,
    costTo: costTo,
    bookingType: values.bookingType,
    bookingOpens: bookingOpens,
    venueId: values.venue.id,
    venueGuidance: values.venueGuidance.trim(),
    eventSeriesId: values.eventSeries.id || null,
    duration: values.duration.trim(),
    useVenueOpeningTimes: values.useVenueOpeningTimes,
    timesRanges: entityMapper.mapSubmittedTimesRanges(values.timesRanges),
    openingTimes: entityMapper.mapSubmittedOpeningTimes(values.openingTimes),
    additionalOpeningTimes: entityMapper.mapSubmittedAdditionalOpeningTimes(
      values.additionalOpeningTimes
    ),
    specialOpeningTimes: entityMapper.mapSubmittedSpecialOpeningTimes(
      values.specialOpeningTimes
    ),
    openingTimesClosures: entityMapper.mapSubmittedOpeningTimesClosures(
      values.openingTimesClosures
    ),
    performances: entityMapper.mapSubmittedPerformances(values.performances),
    additionalPerformances: entityMapper.mapSubmittedAdditionalPerformances(
      values.additionalPerformances
    ),
    specialPerformances: entityMapper.mapSubmittedSpecialPerformances(
      values.specialPerformances
    ),
    performancesClosures: entityMapper.mapSubmittedPerformancesClosures(
      values.performancesClosures
    ),
    talents: entityMapper.mapSubmittedEventTalents(values.talents),
    mediumTags: values.mediumTags,
    styleTags: values.styleTags,
    audienceTags: values.audienceTags,
    geoTags: values.geoTags,
    weSay: (values.weSay || '').trim(),
    rating: parseInt(values.rating),
    minAge: values.minAge ? parseInt(values.minAge) : null,
    // namedClosures: entityMapper.mapSubmittedNamedClosures(values.namedClosures),
    description: entityMapper.mapSubmittedDescription(values.description),
    descriptionCredit: values.descriptionCredit.trim(),
    links: entityMapper.mapSubmittedLinks(values.links),
    images: entityMapper.mapSubmittedImages(values.images),
    version: values.version + 1
  }
}
