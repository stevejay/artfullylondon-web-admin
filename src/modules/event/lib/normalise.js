import * as eventLib from '_src/shared/lib/event'
import * as dateLib from '_src/shared/lib/date'

export function normaliseMoney (value) {
  return value.replace(/[^\d.]/g, '')
}

export function normaliseEventValues (values) {
  const isPerformance = eventLib.eventIsPerformance(values.eventType)
  const isExhibition = !isPerformance

  const isOneTimePerformance =
    isPerformance && eventLib.eventIsOneTime(values.occurrenceType)

  const hasRange = eventLib.occurrenceHasDateRange(values.occurrenceType)

  if (hasRange) {
    if (values.dateFrom > values.dateTo) {
      values.dateTo = values.dateFrom
    }

    values.additionalOpeningTimes = (values.additionalOpeningTimes || [])
      .filter(x => _filterTimeEntries(x, values.dateFrom, values.dateTo))

    values.additionalPerformances = (values.additionalPerformances || [])
      .filter(x => _filterTimeEntries(x, values.dateFrom, values.dateTo))

    values.specialOpeningTimes = (values.specialOpeningTimes || [])
      .filter(x => _filterTimeEntries(x, values.dateFrom, values.dateTo))

    values.specialPerformances = (values.specialPerformances || [])
      .filter(x => _filterTimeEntries(x, values.dateFrom, values.dateTo))

    values.openingTimesClosures = (values.openingTimesClosures || [])
      .filter(x => _filterTimeEntries(x, values.dateFrom, values.dateTo))

    values.performancesClosures = (values.performancesClosures || [])
      .filter(x => _filterTimeEntries(x, values.dateFrom, values.dateTo))
  } else {
    values.dateTo = values.dateFrom = null
  }

  const isSingleDay =
    isOneTimePerformance || (hasRange && values.dateFrom === values.dateTo)

  const isLongerThanWeek =
    hasRange && dateLib.periodIsLongerThanWeek(values.dateFrom, values.dateTo)

  values.useVenueOpeningTimes = isExhibition && values.useVenueOpeningTimes

  if (isExhibition) {
    values.performances = []
    values.additionalPerformances = []
    values.specialPerformances = []
    values.performancesClosures = []

    if (isSingleDay || !isLongerThanWeek || values.useVenueOpeningTimes) {
      values.openingTimes = []
    }

    if (
      isSingleDay ||
      !isLongerThanWeek ||
      values.useVenueOpeningTimes ||
      !hasRange
    ) {
      values.timesRanges = []
    }

    if (isSingleDay) {
      values.openingTimesClosures = []
    }

    if (values.timesRanges && values.timesRanges.length === 0) {
      values.openingTimes = values.openingTimes.map(openingTime => {
        if (openingTime.timesRangeId) {
          delete openingTime.timesRangeId
        }

        return openingTime
      })
    }
  } else {
    values.useVenueOpeningTimes = false
    values.openingTimes = []
    values.additionalOpeningTimes = []
    values.specialOpeningTimes = []
    values.openingTimesClosures = []

    if (isSingleDay || !isLongerThanWeek || isOneTimePerformance) {
      values.performances = []
    }

    if (isSingleDay || !isLongerThanWeek || isOneTimePerformance || !hasRange) {
      values.timesRanges = []
    }

    if (
      isOneTimePerformance &&
      values.additionalPerformances &&
      values.additionalPerformances.length > 1
    ) {
      values.additionalPerformances = values.additionalPerformances.slice(0, 1)
    }

    if (isSingleDay || isOneTimePerformance) {
      values.performancesClosures = []
    }

    if (
      values.timesRanges &&
      values.timesRanges.length === 0 &&
      values.performances
    ) {
      values.performances = values.performances.map(performance => {
        if (performance.timesRangeId) {
          delete performance.timesRangeId
        }

        return performance
      })
    }
  }
}

function _filterTimeEntries (entry, dateFrom, dateTo) {
  return entry.date >= dateFrom && entry.date <= dateTo
}
