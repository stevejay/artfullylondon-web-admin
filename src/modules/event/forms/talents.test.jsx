import React from 'react'
import _ from 'lodash'
import { change } from 'redux-form'

import { EditEventTalentsForm } from './talents'
import * as eventConstants from '../constants'
import entityType from '_src/domain/types/entity-type'
import talentType from '_src/domain/types/talent-type'
import * as arrayLib from '_src/shared/lib/array'
import EntitySelectorSearch from '../components/entity-selector-search'
import Modal from '_src/shared/components/modal'
import Button from '_src/shared/components/button'
import { actions as searchActions } from '_src/modules/search'
import {
  CreateBasicTalentForm,
  actions as talentActions
} from '_src/modules/talent'

it('should render correctly when creating', () => {
  const wrapper = shallow(
    <EditEventTalentsForm
      initialValues={{ name: 'The Name' }}
      isEdit={false}
      submitting={false}
      handleSubmit={func => func}
      onSubmit={_.noop}
      onCancel={_.noop}
      onPreviousPage={_.noop}
      dispatch={_.noop}
      showingModal={false}
      setShowingModal={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should render correctly when editing', () => {
  const wrapper = shallow(
    <EditEventTalentsForm
      initialValues={{ name: 'The Name' }}
      isEdit
      submitting={false}
      handleSubmit={func => func}
      onSubmit={_.noop}
      onCancel={_.noop}
      onPreviousPage={_.noop}
      dispatch={_.noop}
      showingModal={false}
      setShowingModal={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should handle an autocomplete select event', () => {
  arrayLib.addElement = jest.fn().mockReturnValue({ the: 'new value' })
  const dispatch = jest.fn()

  const wrapper = shallow(
    <EditEventTalentsForm
      initialValues={{ name: 'The Name' }}
      isEdit
      submitting={false}
      handleSubmit={func => func}
      onSubmit={_.noop}
      onCancel={_.noop}
      onPreviousPage={_.noop}
      dispatch={dispatch}
      showingModal={false}
      setShowingModal={_.noop}
    />
  )

  wrapper.instance()._talent = { value: 'bar' }

  return wrapper
    .find(EntitySelectorSearch)
    .prop('onAutocompleteSelect')(entityType.TALENT, 'some-talent-id', {
      output: 'The Name',
      status: 'Active',
      talentType: talentType.GROUP,
      commonRole: 'Common Role'
    })
    .then(() => {
      expect(arrayLib.addElement).toHaveBeenCalledWith('bar', {
        key: 'some-talent-id',
        id: 'some-talent-id',
        name: 'The Name',
        status: 'Active',
        talentType: talentType.GROUP,
        commonRole: 'Common Role',
        roles: 'Common Role',
        characters: ''
      })

      expect(dispatch).toHaveBeenCalledWith(
        change(eventConstants.EDIT_EVENT_TALENTS_FORM_NAME, 'talents', {
          the: 'new value'
        })
      )
    })
})

it('should handle an autocomplete select event when the talent already exists', () => {
  arrayLib.addElement = jest.fn().mockReturnValue(null)
  const dispatch = jest.fn()

  const wrapper = shallow(
    <EditEventTalentsForm
      initialValues={{ name: 'The Name' }}
      isEdit
      submitting={false}
      handleSubmit={func => func}
      onSubmit={_.noop}
      onCancel={_.noop}
      onPreviousPage={_.noop}
      dispatch={dispatch}
      showingModal={false}
      setShowingModal={_.noop}
    />
  )

  wrapper.instance()._talent = { value: 'bar' }

  return wrapper
    .find(EntitySelectorSearch)
    .prop('onAutocompleteSelect')(entityType.TALENT, 'some-talent-id', {
      output: 'The Name',
      status: 'Active',
      talentType: talentType.GROUP,
      commonRole: 'Common Role'
    })
    .then(
      () => {
        throw new Error('test failed')
      },
      () => {
        expect(dispatch).not.toHaveBeenCalled()
      }
    )
})

it('should handle an autocomplete search event', () => {
  const dispatch = jest.fn().mockReturnValue(Promise.resolve())

  const wrapper = shallow(
    <EditEventTalentsForm
      initialValues={{ name: 'The Name' }}
      isEdit
      submitting={false}
      handleSubmit={func => func}
      onSubmit={_.noop}
      onCancel={_.noop}
      onPreviousPage={_.noop}
      dispatch={dispatch}
      showingModal={false}
      setShowingModal={_.noop}
    />
  )

  return wrapper
    .find(EntitySelectorSearch)
    .prop('onAutocompleteSearch')('some term', entityType.TALENT)
    .then(() => {
      expect(dispatch).toHaveBeenCalledWith(
        searchActions.autocompleteSearch('some term', entityType.TALENT)
      )
    })
})

it('should handle showing the create talent modal', () => {
  const setShowingModal = jest.fn()

  const wrapper = shallow(
    <EditEventTalentsForm
      initialValues={{ name: 'The Name' }}
      isEdit
      submitting={false}
      handleSubmit={func => func}
      onSubmit={_.noop}
      onCancel={_.noop}
      onPreviousPage={_.noop}
      dispatch={_.noop}
      showingModal={false}
      setShowingModal={setShowingModal}
    />
  )

  wrapper.find(Button).simulate('click')

  expect(setShowingModal).toHaveBeenCalledWith(true)
})

it('should handle hiding the create talent modal', () => {
  const setShowingModal = jest.fn()

  const wrapper = shallow(
    <EditEventTalentsForm
      initialValues={{ name: 'The Name' }}
      isEdit
      submitting={false}
      handleSubmit={func => func}
      onSubmit={_.noop}
      onCancel={_.noop}
      onPreviousPage={_.noop}
      dispatch={_.noop}
      showingModal
      setShowingModal={setShowingModal}
    />
  )

  wrapper.find(Modal).simulate('hide')

  expect(setShowingModal).toHaveBeenCalledWith(false)
})

it('should handle a create talent event', () => {
  const setShowingModal = jest.fn()
  const dispatch = jest.fn().mockReturnValue(Promise.resolve())

  const wrapper = shallow(
    <EditEventTalentsForm
      initialValues={{ name: 'The Name' }}
      isEdit
      submitting={false}
      handleSubmit={func => func}
      onSubmit={_.noop}
      onCancel={_.noop}
      onPreviousPage={_.noop}
      dispatch={dispatch}
      showingModal
      setShowingModal={setShowingModal}
    />
  )

  return wrapper
    .find(CreateBasicTalentForm)
    .prop('onSubmit')({ the: 'values' })
    .then(() => {
      expect(dispatch).toHaveBeenCalledWith(
        talentActions.createTalentForEntity(
          { the: 'values' },
          eventConstants.EDIT_EVENT_TALENTS_FORM_NAME
        )
      )

      expect(setShowingModal).toHaveBeenCalledWith(false)
    })
})
