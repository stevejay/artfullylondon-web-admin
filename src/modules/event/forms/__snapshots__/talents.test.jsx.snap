// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly when creating 1`] = `
<React.Fragment>
  <Form
    onSubmit={[Function]}
  >
    <FormSectionHeader>
      Talents
    </FormSectionHeader>
    <withState(EntitySelectorSearch)
      entityType="talent"
      onAutocompleteSearch={[Function]}
      onAutocompleteSelect={[Function]}
    />
    <FormRow>
      <Field
        component={[Function]}
        containerStyle={
          Object {
            "paddingTop": 0,
          }
        }
        label=""
        name="talents"
      />
    </FormRow>
    <Button
      aria-label="Create a new Talent"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Create a new Talent…
    </Button>
    <Divider
      className="src-modules-event-forms-talents__last-divider"
      type="basic"
    />
    <FormError />
    <FormButtons
      onCancel={null}
      onPrevious={[Function]}
      submitLabel="Next"
      submitting={false}
    />
  </Form>
  <Modal
    aria-label="Datepicker"
    backdropTransition={[Function]}
    onHide={[Function]}
    show={false}
    transition={[Function]}
  >
    <ModalContainer
      dismissable={true}
      onHide={[Function]}
      title="Create New Talent"
      type="wide"
    >
      <Connect(ReduxForm)
        className="src-modules-event-forms-talents__modal-form"
        initialValues={
          Object {
            "commonRole": "",
            "firstNames": "",
            "images": Array [],
            "lastName": "",
            "links": Array [],
            "status": "Active",
            "talentType": "Individual",
            "version": 0,
            "weSay": "",
          }
        }
        onCancel={[Function]}
        onSubmit={[Function]}
      />
    </ModalContainer>
  </Modal>
</React.Fragment>
`;

exports[`should render correctly when editing 1`] = `
<React.Fragment>
  <Form
    onSubmit={[Function]}
  >
    <FormSectionHeader>
      Talents
    </FormSectionHeader>
    <withState(EntitySelectorSearch)
      entityType="talent"
      onAutocompleteSearch={[Function]}
      onAutocompleteSelect={[Function]}
    />
    <FormRow>
      <Field
        component={[Function]}
        containerStyle={
          Object {
            "paddingTop": 0,
          }
        }
        label=""
        name="talents"
      />
    </FormRow>
    <Button
      aria-label="Create a new Talent"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Create a new Talent…
    </Button>
    <Divider
      className="src-modules-event-forms-talents__last-divider"
      type="basic"
    />
    <FormError />
    <FormButtons
      onCancel={[Function]}
      onPrevious={[Function]}
      submitLabel="Next"
      submitting={false}
    />
  </Form>
  <Modal
    aria-label="Datepicker"
    backdropTransition={[Function]}
    onHide={[Function]}
    show={false}
    transition={[Function]}
  >
    <ModalContainer
      dismissable={true}
      onHide={[Function]}
      title="Create New Talent"
      type="wide"
    >
      <Connect(ReduxForm)
        className="src-modules-event-forms-talents__modal-form"
        initialValues={
          Object {
            "commonRole": "",
            "firstNames": "",
            "images": Array [],
            "lastName": "",
            "links": Array [],
            "status": "Active",
            "talentType": "Individual",
            "version": 0,
            "weSay": "",
          }
        }
        onCancel={[Function]}
        onSubmit={[Function]}
      />
    </ModalContainer>
  </Modal>
</React.Fragment>
`;
